cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -308,19 +319,89 @@ TEST_F(TestExtensionType, TestFromTensorType) {
   auto dim_names = std::vector<std::vector<std::string>>{
       {"y", "z"}, {"z", "y"}, {"y", "z"}, {"z", "y"},
       {"y", "z"}, {"y", "z"}, {"y", "z"}, {"y", "z"}};
-  auto cell_shapes = std::vector<std::vector<int64_t>>{{3, 4}, {4, 3}, {4, 3}, {3, 4}};
+  auto element_shapes = std::vector<std::vector<int64_t>>{{3, 4}, {4, 3}, {4, 3}, {3, 4}};
   auto permutations = std::vector<std::vector<int64_t>>{{0, 1}, {1, 0}, {0, 1}, {1, 0}};
 
   for (size_t i = 0; i < shapes.size(); i++) {
     ASSERT_OK_AND_ASSIGN(auto tensor, Tensor::Make(value_type_, values, shapes[i],
                                                    strides[i], tensor_dim_names[i]));
     ASSERT_OK_AND_ASSIGN(auto ext_arr, FixedShapeTensorArray::FromTensor(tensor));
     auto ext_type =
-        fixed_shape_tensor(value_type_, cell_shapes[i], permutations[i], dim_names[i]);
+        fixed_shape_tensor(value_type_, element_shapes[i], permutations[i], dim_names[i]);
     CheckFromTensorType(tensor, ext_type);
   }
 }
 
+template <typename T>
+void CheckToTensor(const std::vector<T>& values, const std::shared_ptr<DataType> typ,
+                   const int32_t& element_size, const std::vector<int64_t>& element_shape,
+                   const std::vector<int64_t>& element_permutation,
+                   const std::vector<std::string>& element_dim_names,
+                   const std::vector<int64_t>& tensor_shape,
+                   const std::vector<std::string>& tensor_dim_names,
+                   const std::vector<int64_t>& tensor_strides) {
+  auto buffer = Buffer::Wrap(values);
+  const std::shared_ptr<DataType> element_type = fixed_size_list(typ, element_size);
+  std::vector<std::shared_ptr<Buffer>> buffers = {nullptr, buffer};
+  auto arr_data = std::make_shared<ArrayData>(typ, values.size(), buffers);
+  auto arr = std::make_shared<Int64Array>(arr_data);
+  ASSERT_OK_AND_ASSIGN(auto fsla_arr, FixedSizeListArray::FromArrays(arr, element_type));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto expected_tensor,
+      Tensor::Make(typ, buffer, tensor_shape, tensor_strides, tensor_dim_names));
+  const auto ext_type =
+      fixed_shape_tensor(typ, element_shape, element_permutation, element_dim_names);
+
+  auto ext_arr = ExtensionType::WrapArray(ext_type, fsla_arr);
+  const auto tensor_array = std::static_pointer_cast<FixedShapeTensorArray>(ext_arr);
+  ASSERT_OK_AND_ASSIGN(const auto actual_tensor, tensor_array->ToTensor());
```

Comment:
Could you perhaps add validation here? Something like:
```suggestion
  ASSERT_OK_AND_ASSIGN(const auto actual_tensor, tensor_array->ToTensor());
  ASSERT_OK(actual_tensor->Validate());
```